### ISSUE TITLE
Content Security Policy (CSP) Bypass via Same-Origin iFrames

### ISSUE THREAD

**Issue by RedYetiDev** [w3c/webappsec-csp#652]
## Issue
The current Content Security Policy (CSP) specification allows for script exploitation via same-origin iFrames, undermining its intended function of mitigating cross-site scripting (XSS) attacks. While CSP aims to restrict script execution and other potentially harmful content, this vulnerability enables scripts within same-origin iFrames to execute within the context of the parent page, effectively circumventing CSP directives.

## Impact
Exploiting this vulnerability presents severe consequences:
- **XSS Attacks**: Malicious scripts executed within the parent page's context can manipulate user sessions, steal sensitive data, or perform unauthorized actions.
- **Content Injection**: Attackers can inject deceptive content, like phishing forms or fraudulent advertisements, into the parent page, jeopardizing user trust and potentially leading to further exploitation.
- **Data Exfiltration**: By exploiting CSP bypass, attackers can extract sensitive user data from the parent page to external servers, breaching privacy regulations and compromising user security.

## Technical Overview
The vulnerability stems from how CSP directives are enforced within same-origin iFrames. While CSP effectively limits script execution from external sources on the parent page, it does not prevent scripts within same-origin iFrames from executing within the parent page's context. This enables attackers to bypass CSP restrictions by embedding malicious scripts within iFrames hosted on the same origin as the parent page.

#### Example
Consider a web application implementing a strict CSP policy to prevent XSS attacks:
```
Content-Security-Policy: default-src 'self'; script-src 'self';
```

Despite this policy, an attacker can embed a same-origin iFrame within the application's page:
```html
<script>window.secret = "s3cr3t"</script>
<iframe src="https://example.com/"></iframe>
```
Within this iFrame, the attacker injects a malicious script:
```html
<script>
console.log(window.top.secret)
</script>
```
The issue lies not in the script itself but in the fact that this script, though restricted by the CSP policy if loaded directly onto the parent page, executes within the parent page's context when loaded within the iFrame, effectively bypassing CSP restrictions.

## Resolution vs `sandbox`
To address the CSP issue effectively, I propose implementing a more comprehensive specification that delineates precisely when iFrames are permitted or prohibited from executing specific tasks, such as scripting. This approach aims to enhance the control and granularity of CSP directives, ensuring tighter security measures compared to the existing `sandbox` attribute. 

Here's why this approach is more effective:

1. **Fine-Grained Control**: Unlike the `sandbox` attribute, which applies a blanket set of restrictions to all content within the iFrame, a more robust CSP specification allows for nuanced control over the behavior of scripts within iFrames. By specifying precisely when and under what conditions scripts can execute, administrators can tailor security policies to match the specific requirements of their applications.

2. **Compatibility and Flexibility**: While the `sandbox` attribute imposes uniform restrictions on all content within an iFrame, a more detailed CSP specification offers greater flexibility and compatibility. It enables developers to implement security measures that are tailored to the unique characteristics and functionalities of their web applications, without resorting to overly restrictive or cumbersome restrictions.

3. **Enhanced Security Posture**: By explicitly defining the circumstances under which scripts can execute within iFrames, the proposed CSP specification provides a more robust defense against malicious exploitation. Administrators can mitigate the risk of XSS attacks and other security vulnerabilities by tightly controlling script execution, thereby bolstering the overall security posture of their web applications.

In summary, implementing a more comprehensive CSP specification that delineates the conditions under which iFrames can perform scripting tasks offers several advantages over the `sandbox` attribute. It provides finer control, greater compatibility and flexibility, enhanced security posture, and clearer enforcement mechanisms, thereby strengthening the overall security of web applications.

## Proof Of Concept
[poc.tar.gz](https://github.com/w3c/webappsec-csp/files/14781500/poc.tar.gz): If hosted, `s3cr3t` will be `console.log`ed, despite it being declared in a situation where CSP is set to only `allow-self`

**Comment by RedYetiDev** [comment]
As I've heard, this feature would be too confusing to implement, and the suggested alternative is sandboxing.
